[RFC]: Create aten native op for constrain_range

XuanQi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103346

At high current implementation of constrains functions (constrain_as_**) will raise exception for the following code snippets:
```
def f(x):
    a = x.item()
    constrain_as_size(a, 4, 7)
    return torch.empty((a, 4))

inp = torch.tensor([5])
ep = torch._export.export(f, (inp,))
```

The reason is because current constrain logic is:
1) Purely python so it won't survive AOT export (the full node is gone after AOT export since AOT export only maintains aten level op).
2) Utilize side effect to add range constraints for traced symbol's shape env ([code](https://github.com/pytorch/pytorch/blob/9591e52880fb09cb6753d97606e6efb956075f5b/torch/fx/experimental/symbolic_shapes.py#L370-L372)).
3) If runtime assertion is turned on (by default). [`_AddRuntimeAssertionsForConstraintsPass`](https://github.com/pytorch/pytorch/blob/9591e52880fb09cb6753d97606e6efb956075f5b/torch/_export/passes/add_runtime_assertions_for_constraints_pass.py#L98-L100) will try to append assertion node based on range constrains extracted from shape env of symbol during another interpretation round.
4). However, since 1), in the round of AOT export, range constraints logic won't run for symbols generated during this round. And later there is no range constrains information available for assertion round and caused issue.
5) As a result of above, it will failure at `torch.empty((a, 4))` (there is no constrains for `a` that it must be positive).

The fix here is just to implement range constrain logic as a native aten op (CPU implementation as no-op) to make it be able to survive AOT export.

**NOTE:**
[Logic](https://github.com/pytorch/pytorch/blob/2d745b95d723641e575027bd4e2fff612f61cc8f/torch/fx/experimental/symbolic_shapes.py#L350-L365C15) within [`constrain_range`](https://github.com/pytorch/pytorch/blob/2d745b95d723641e575027bd4e2fff612f61cc8f/torch/fx/experimental/symbolic_shapes.py#LL313C74-L313C74) is split out as `constrain_range_int` to capture case when non `SymInt` is passed in and reused in the new `_constrain_range`. The reason is when non `SymInt` is provided:
* If it directly calls `sym_constrain_range`, the C++ version will be called which will be no-op.
* So in this case it calls `constrain_range_int` instead to be able to capture issue like user provides a input whose tensor's shape could be out of range during exporting, like the following for above code example:
```
...
inp = torch.tensor([10])
ep = torch._export.export(f, (inp,)) # immediately raise error
```

Differential Revision: [D46734204](https://our.internmc.facebook.com/intern/diff/D46734204)

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78

XuanQi committed:
[RFC]: Create aten native op for constrain_range

[ghstack-poisoned]

XuanQi commented:
@xuanqi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D46611045).

XuanQi committed:
Update on "[RFC]: Create aten native op for constrain_range"




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy aakhundov

Differential Revision: [D46611045](https://our.internmc.facebook.com/intern/diff/D46611045)

[ghstack-poisoned]

XuanQi commented:
@xuanqi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D46611045).

XuanQi commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -177,6 +177,10 @@
 
 - func: _assert_tensor_metadata(Tensor a, SymInt[]? size=None, SymInt[]? stride=None, ScalarType? dtype=None) -> ()
 
+- func: _constrain_range_native(Scalar size, int? min=None, int? max=None) -> ()
```
Do we want to keep the leading `_` or just `sym_constrain_range`?

XuanQi committed:
Update on "[RFC]: Create aten native op for constrain_range"




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy aakhundov

Differential Revision: [D46611045](https://our.internmc.facebook.com/intern/diff/D46611045)

[ghstack-poisoned]

XuanQi commented on torch/_export/constraints.py:
```diff
@@ -1,10 +1,23 @@
+        constrain_range_non_symint(symbol, min=min, max=max)
+        return
+
+    torch.sym_constrain_range(symbol, min, max)
```
@tugsbayasgalan: do we actually want to manually suppress this? I didn't see there is an Sym* version of Scalar so I guess the way to support this is to have overloads for int (SymInt in native_functions.yaml) or float (SymFloat in native_functions.yaml)?

XuanQi committed:
Update on "[RFC]: Create aten native op for constrain_range"




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy aakhundov

Differential Revision: [D46611045](https://our.internmc.facebook.com/intern/diff/D46611045)

[ghstack-poisoned]

XuanQi commented:
@xuanqi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D46611045).

eellison commented:
Can you add a commit summary?

XuanQi committed:
Update on "[RFC]: Create aten native op for constrain_range"


At high since current implementation of constrains functions (constrain_as_**) will raise exception for the following code snippets:
```
def f(x):
    a = x.item()
    constrain_as_size(a, 4, 7)
    return torch.empty((a, 4))

inp = torch.tensor([5])
ep = torch._export.export(f, (inp,))
```

The reason is because current constrain logic is:
1) Purely python so it won't survive AOT export (the full node is gone after AOT export since AOT export only maintains aten level op).
2) Utilize side effect to add range constraints for traced symbol's shape env ([code](https://github.com/pytorch/pytorch/blob/9591e52880fb09cb6753d97606e6efb956075f5b/torch/fx/experimental/symbolic_shapes.py#L370-L372)).
3) If runtime assertion is turned on (by default). [`_AddRuntimeAssertionsForConstraintsPass`](https://github.com/pytorch/pytorch/blob/9591e52880fb09cb6753d97606e6efb956075f5b/torch/_export/passes/add_runtime_assertions_for_constraints_pass.py#L98-L100) will try to append assertion node based on range constrains extracted from shape env of symbol during another interpretation round.
4). However, since 1), in the round of AOT export, range constraints logic won't run for symbols generated during this round. And later there is no range constrains information available for assertion round and caused issue.

The fix here is just to implement range constrain logic as a native aten op (CPU implementation as no-op) to make it be able to survive AOT export.

Differential Revision: [D46611045](https://our.internmc.facebook.com/intern/diff/D46611045)

[ghstack-poisoned]

XuanQi commented:
@xuanqi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D46611045).

XuanQi committed:
Update on "[RFC]: Create aten native op for constrain_range"


At high current implementation of constrains functions (constrain_as_**) will raise exception for the following code snippets:
```
def f(x):
    a = x.item()
    constrain_as_size(a, 4, 7)
    return torch.empty((a, 4))

inp = torch.tensor([5])
ep = torch._export.export(f, (inp,))
```

The reason is because current constrain logic is:
1) Purely python so it won't survive AOT export (the full node is gone after AOT export since AOT export only maintains aten level op).
2) Utilize side effect to add range constraints for traced symbol's shape env ([code](https://github.com/pytorch/pytorch/blob/9591e52880fb09cb6753d97606e6efb956075f5b/torch/fx/experimental/symbolic_shapes.py#L370-L372)).
3) If runtime assertion is turned on (by default). [`_AddRuntimeAssertionsForConstraintsPass`](https://github.com/pytorch/pytorch/blob/9591e52880fb09cb6753d97606e6efb956075f5b/torch/_export/passes/add_runtime_assertions_for_constraints_pass.py#L98-L100) will try to append assertion node based on range constrains extracted from shape env of symbol during another interpretation round.
4). However, since 1), in the round of AOT export, range constraints logic won't run for symbols generated during this round. And later there is no range constrains information available for assertion round and caused issue.
5) As a result of above, it will failure at `torch.empty((a, 4))` (there is no constrains for `a` that it must be positive).

The fix here is just to implement range constrain logic as a native aten op (CPU implementation as no-op) to make it be able to survive AOT export.

Differential Revision: [D46611045](https://our.internmc.facebook.com/intern/diff/D46611045)

[ghstack-poisoned]

avikchaudhuri commented on test/dynamo/test_export.py:
```diff
@@ -2411,6 +2411,29 @@ def f(x, y):
+
+            ep = torch._export.export(f, (torch.tensor([5]),))
+            self.assertEqual(
+                ep.graph_module(torch.tensor([6]))[0],
```
Should be able to just call `ep` here and below.

avikchaudhuri commented on torch/_export/constraints.py:
```diff
@@ -1,9 +1,31 @@
+# could cause type error during since `SymInt` or `SymFloat` will be used.
+# Here manually specify the type explicitly.
+sym_constrain_range: Callable[
+    [Union[Number, _complex, SymInt, SymFloat], Optional[int], Optional[int]],
```
Why accept `_complex`? Also, maybe `int, float` instead of `Number`?

XuanQi replied:
As mentioned in the comment `Scalar` is translated to python typing as `Union[Number, _complex]` but yeah we could keep it simple and only include `int` and `float` since we're redefining the type anyway.

avikchaudhuri commented on torch/_export/constraints.py:
```diff
@@ -1,9 +1,31 @@
+        # an value derived from numeric input) for exporting is within range or
+        # not.
+        constrain_range_non_symint(symbol, min=min, max=max)
+        return
```
Just use `else` instead of plain `return`?

avikchaudhuri replied:
Umm, actually I don't follow the logic here. Should you make `constrain_range_non_symint` return a symint that you can then call `sym_constraint_range` on?

avikchaudhuri replied:
A satisfactory way to resolve this would be to write a test that actually exercises this path.

XuanQi replied:
Use the following logic as example:
```
def f(x):
    a = x.item()
    constrain_as_size(a, 4, 7)
    return torch.empty((a, 4))
```
The goal here is that if user passes in tensor like `torch.tensor([20])` during export, it will immediately fail since it's out of range. And based on my observation there is some round (I think it's the first round) of parsing, `20` (as `int`) will be passed in as `symbol` and the purpose of this logic is to test on this case.

> Should you make constrain_range_non_symint return a symint that you can then call sym_constraint_range on
Can try it out.

> A satisfactory way to resolve this would be to write a test that actually exercises this path.
This case is tested in the newly added test (`torch._export.export(f, (torch.tensor([20]),))`)

avikchaudhuri commented on torch/_export/constraints.py:
```diff
@@ -1,9 +1,31 @@
+        constrain_range_non_symint(symbol, min=min, max=max)
+        return
+
+    sym_constrain_range(symbol, min, max)
```
So you're only ever going to call this if you get a `SymInt`?

XuanQi replied:
As mentioned above yes. This indeed a little bit strange though but seems like the meta implementation is not triggered when real value is passed in though. I'm not sure if the real c++ implementation is triggered in this case though but can check.

avikchaudhuri commented on torch/_export/passes/add_runtime_assertions_for_constraints_pass.py:
```diff
@@ -60,8 +60,23 @@ def __init__(
+        # 2. b.shape[0] passed to constrain_as_value(...).
+        # NOTE: cached_assertions will cause _AddRuntimeAssertionsForConstraintsPass stateful
+        # but only within each pass - it will always be reset at start of each pass.
+        self.cached_assertions: Set[Symbol] = set()
```
What's the problem you're trying to solve?

1. Are duplicate assertions a problem you must solve in this diff?
2. If we want to deduplicate assertions, let's do it for all rather than just for inline constraints.

XuanQi replied:
Just to fix a failed test case https://github.com/pytorch/pytorch/blob/6cc0f1c20c2f87a6c7b0e4abd5419e5007920999/test/export/test_passes.py#L227.
But anyway, yeah it might be better to handle this separately and I guess we also need to handle case https://github.com/pytorch/pytorch/blob/93b0410eef57ebf038c12ed2fa1d4018a24096b7/torch/_export/passes/add_runtime_assertions_for_constraints_pass.py#L101-L106.

I will revert this change and relax the test case for now and leave a TODO to handle duplication. cc @tugsbayasgalan  for awareness.

avikchaudhuri replied:
What were the counts in that test before caching?

avikchaudhuri replied:
I'm a bit uncomfortable not understanding what change in this diff caused the counts to change in that test, and therefore required caching. Specifically, are we doing something in this diff that *causes* the same assertion to be added twice, and can we avoid it there rather than doing caching here?

XuanQi replied:
So use the [test example](https://github.com/pytorch/pytorch/blob/a60f6dbe69b0cae2e27ac279cdf399c6c16650af/test/export/test_passes.py#L227), I think TLDR is that the second assertion comes from `constrain_as_value` survived AOT export.
More specifically:
Before the change and after AOT export, the logic is roughly like:
```
b = x.nonzero() # <- the assertion is for `b.shape[0]` here (it can smartly verify dimension of `b` because the recursion logic at https://fburl.com/bw1gihkx)
# constrain_as_value(b.shape[0], min=3, max=5) <- this statement is totally gone.
return b
```
After the change and after AOT export, the logic is roughly like:
```
b = x.nonzero() # <- the assertion is for `b.shape[0]` as before.
constrain_as_value(b.shape[0], min=3, max=5), min=2, max=5) # <- the new assertion is for `b.shape[0]` inside `constrain_as_value(...)`
```
Since _AddRuntimeAssertionsForConstraintsPass runs in interpretation mode, which means each statement is executed sequentially, so the the same assertion is added again when executing `b.shape[0]` which is the input of `constrain_as_value`. So the fix here is just to add a cache to avoid adding assertion for the same symbol twice.

avikchaudhuri replied:
I see. @tugsbayasgalan @angelayi any thoughts here on whether this is expected behavior? How are we producing an assertion at `b = x.nonzero()`, and if so, how does it know the bounds are `(3,5)` without looking at the next line? Is it because in an earlier pass we discovered the constraint to add, so in the next pass we're adding it again? If so, can't we check the shape env or something to avoid adding it again?

To be clear, I have nothing against doing caching to avoid this duplication. What I'm actually worried about is that this is hiding some bug, or at least some ugliness, in our logic. Please let me know if I'm wrong.

XuanQi replied:
Also did another sanity check, for the follow logic **before the PR**:
```
class M(torch.nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, x):
        b = x.nonzero()
        constrain_as_value(b.shape[0], min=3, max=5)
        z = b.shape[0] + 1
        return z

mod = M()
x = torch.tensor([2, 1, 2, 3, 5, 0])
ep = export(mod, (x,), constraints=[dynamic_dim(x, 0) >= 2])

print(count_call_function(ep.graph, torch.ops.aten._assert_async.msg)) # this will be 4 as well.
```
So I think this verifies this is not a new behavior introduced by this PR :p

tugsbayasgalan replied:
Could you paste the whole graph before your change here?

XuanQi replied:
So the graph (before my change) for the code snippets I've posted above is:
```graph():
    %arg0_1 : [#users=2] = placeholder[target=arg0_1]
    %sym_size_int : [#users=0] = call_function[target=torch.ops.aten.sym_size.int](args = (%arg0_1, 0), kwargs = {})
    %nonzero_default : [#users=2] = call_function[target=torch.ops.aten.nonzero.default](args = (%arg0_1,), kwargs = {})
    %sym_size_int_1 : [#users=2] = call_function[target=torch.ops.aten.sym_size.int](args = (%nonzero_default, 0), kwargs = {})
    %ge : [#users=1] = call_function[target=operator.ge](args = (%sym_size_int_1, 3), kwargs = {})
    %scalar_tensor_default : [#users=1] = call_function[target=torch.ops.aten.scalar_tensor.default](args = (%ge,), kwargs = {})
    %_assert_async_msg : [#users=0] = call_function[target=torch.ops.aten._assert_async.msg](args = (%scalar_tensor_default, nonzero_default.shape[0] is outside of inline constraint [3, 5].), kwargs = {})
    %le : [#users=1] = call_function[target=operator.le](args = (%sym_size_int_1, 5), kwargs = {})
    %scalar_tensor_default_1 : [#users=1] = call_function[target=torch.ops.aten.scalar_tensor.default](args = (%le,), kwargs = {})
    %_assert_async_msg_1 : [#users=0] = call_function[target=torch.ops.aten._assert_async.msg](args = (%scalar_tensor_default_1, nonzero_default.shape[0] is outside of inline constraint [3, 5].), kwargs = {})
    %sym_size : [#users=3] = call_function[target=torch.ops.aten.sym_size](args = (%nonzero_default, 0), kwargs = {})
    %ge_1 : [#users=1] = call_function[target=operator.ge](args = (%sym_size, 3), kwargs = {})
    %scalar_tensor_default_2 : [#users=1] = call_function[target=torch.ops.aten.scalar_tensor.default](args = (%ge_1,), kwargs = {})
    %_assert_async_msg_2 : [#users=0] = call_function[target=torch.ops.aten._assert_async.msg](args = (%scalar_tensor_default_2, sym_size is outside of inline constraint [3, 5].), kwargs = {})
    %le_1 : [#users=1] = call_function[target=operator.le](args = (%sym_size, 5), kwargs = {})
    %scalar_tensor_default_3 : [#users=1] = call_function[target=torch.ops.aten.scalar_tensor.default](args = (%le_1,), kwargs = {})
    %_assert_async_msg_3 : [#users=0] = call_function[target=torch.ops.aten._assert_async.msg](args = (%scalar_tensor_default_3, sym_size is outside of inline constraint [3, 5].), kwargs = {})
    %add : [#users=1] = call_function[target=operator.add](args = (%sym_size, 1), kwargs = {})
    return (add,)
```

Actually I feel like there is another issue? Should there be a constrains for `dynamic_dim(x, 0) >= 2`? Maybe I missed it or misunderstand things :p.

angelayi replied:
Hmm, yes seems like a bug 🤔

angelayi replied:
Oh wait sorry, it's not a bug, because we always want the shapes to be >= 2.

avikchaudhuri commented on torch/_export/passes/add_runtime_assertions_for_constraints_pass.py:
```diff
@@ -254,7 +269,11 @@ def sym_size_cb(proxy, assert_msg, dim):
 
-    def _assert_range_constraint(self, proxy, lower, upper, assert_msg):
+    def _assert_range_constraint(self, *, proxy, lower, upper, assert_msg, symbol):
+        if symbol in self.cached_assertions:
```
Are we sure we will only have one `(lower, upper)` per symbol?

XuanQi replied:
Currently `self.range_constraints` is map from symbol to `RangeConstraint` so there is at mostly one (lower, upper) tuple per symbol? Let me add some comment to make this more explicit.

avikchaudhuri replied:
OK.

avikchaudhuri commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -348,6 +336,24 @@ def constrain_range(a, *, min: Optional[int], max: Optional[int] = None):
+        shape_env.var_to_range[sym_integer] = ValueRanges(min, max)
+        shape_env.var_to_stack[sym_integer] = TracingContext(fake_mode).extract_stack()
+
+    return
```
delete

XuanQi replied:
Let me follow up with @tugsbayasgalan and @angelayi for some context of this logic.

avikchaudhuri replied:
OK, thanks. Just to be clear, here where I said delete I just mean the `return`, because it's superfluous. But yeah, good idea to get some context...I'm not sure myself when this or the other code path is taken (something is a symint vs. not).

avikchaudhuri commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -348,6 +336,24 @@ def constrain_range(a, *, min: Optional[int], max: Optional[int] = None):
         builtins.max(r.lower, min), builtins.min(r.upper, max)
     )
 
+def constrain_range_non_symint(a, *, min, max):
```
This is a strange function, feels like it should be private.

Also, the name is misleading, it's not really constraining anything, but rather creating a symint.

Should you be returning that symint? In that case, you can rename the function to be something like `create_*` or `convert_*`.

angelayi replied:
Context for that code is in https://github.com/pytorch/pytorch/pull/101655, but I'll leave it up to Xuan on how he wants to separate this out into a function.

XuanQi replied:
How about we rename it to something like `constrain_range_eager`?

avikchaudhuri replied:
I still don't like the name. Something like `check_range_and_create_symbol` would be more appropriate.

tugsbayasgalan replied:
Agreed with @avikchaudhuri . I think `constrain_range_eager` is little vague. `check_range_and_create_symbol` sounds better.

avikchaudhuri reviewed (dismissed):
Requesting changes mostly because I'm not sure what's going on with non-symint case.

XuanQi committed:
Update on "[RFC]: Create aten native op for constrain_range"


At high current implementation of constrains functions (constrain_as_**) will raise exception for the following code snippets:
```
def f(x):
    a = x.item()
    constrain_as_size(a, 4, 7)
    return torch.empty((a, 4))

inp = torch.tensor([5])
ep = torch._export.export(f, (inp,))
```

The reason is because current constrain logic is:
1) Purely python so it won't survive AOT export (the full node is gone after AOT export since AOT export only maintains aten level op).
2) Utilize side effect to add range constraints for traced symbol's shape env ([code](https://github.com/pytorch/pytorch/blob/9591e52880fb09cb6753d97606e6efb956075f5b/torch/fx/experimental/symbolic_shapes.py#L370-L372)).
3) If runtime assertion is turned on (by default). [`_AddRuntimeAssertionsForConstraintsPass`](https://github.com/pytorch/pytorch/blob/9591e52880fb09cb6753d97606e6efb956075f5b/torch/_export/passes/add_runtime_assertions_for_constraints_pass.py#L98-L100) will try to append assertion node based on range constrains extracted from shape env of symbol during another interpretation round.
4). However, since 1), in the round of AOT export, range constraints logic won't run for symbols generated during this round. And later there is no range constrains information available for assertion round and caused issue.
5) As a result of above, it will failure at `torch.empty((a, 4))` (there is no constrains for `a` that it must be positive).

The fix here is just to implement range constrain logic as a native aten op (CPU implementation as no-op) to make it be able to survive AOT export.

Differential Revision: [D46611045](https://our.internmc.facebook.com/intern/diff/D46611045)

[ghstack-poisoned]

XuanQi committed:
Update on "[RFC]: Create aten native op for constrain_range"


At high current implementation of constrains functions (constrain_as_**) will raise exception for the following code snippets:
```
def f(x):
    a = x.item()
    constrain_as_size(a, 4, 7)
    return torch.empty((a, 4))

inp = torch.tensor([5])
ep = torch._export.export(f, (inp,))
```

The reason is because current constrain logic is:
1) Purely python so it won't survive AOT export (the full node is gone after AOT export since AOT export only maintains aten level op).
2) Utilize side effect to add range constraints for traced symbol's shape env ([code](https://github.com/pytorch/pytorch/blob/9591e52880fb09cb6753d97606e6efb956075f5b/torch/fx/experimental/symbolic_shapes.py#L370-L372)).
3) If runtime assertion is turned on (by default). [`_AddRuntimeAssertionsForConstraintsPass`](https://github.com/pytorch/pytorch/blob/9591e52880fb09cb6753d97606e6efb956075f5b/torch/_export/passes/add_runtime_assertions_for_constraints_pass.py#L98-L100) will try to append assertion node based on range constrains extracted from shape env of symbol during another interpretation round.
4). However, since 1), in the round of AOT export, range constraints logic won't run for symbols generated during this round. And later there is no range constrains information available for assertion round and caused issue.
5) As a result of above, it will failure at `torch.empty((a, 4))` (there is no constrains for `a` that it must be positive).

The fix here is just to implement range constrain logic as a native aten op (CPU implementation as no-op) to make it be able to survive AOT export.

Differential Revision: [D46611045](https://our.internmc.facebook.com/intern/diff/D46611045)

[ghstack-poisoned]

XuanQi commented:
@xuanqi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D46734204).

tugsbayasgalan commented on torch/_export/constraints.py:
```diff
@@ -13,7 +30,7 @@ def constrain_as_value(symbol, min: Optional[int] = None, max: Optional[int] = N
     """
 
-    constrain_range(symbol, min=min, max=max)
+    _constrain_range(symbol, min=min, max=max)
```
Why not just inline `_constrain_range`?

XuanQi replied:
Will do. Mostly some outcome of changes between revisions but yeah it makes sense to make it inline.

tugsbayasgalan commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -348,6 +336,22 @@ def constrain_range(a, *, min: Optional[int], max: Optional[int] = None):
         builtins.max(r.lower, min), builtins.min(r.upper, max)
     )
 
+def constrain_range_eager(a, *, min, max):
```
nit: Could you add a comment mentioning that this code path is run in eager so we can't just error whenever we see a concrete integer?

tugsbayasgalan replied:
Basically what we discussed yesterday. It seemed weird that we were modifying shape_env for concrete integers, but the real reason turned out that we want to be able to execute this op in eager mode, hence we can't just error whenever we see concrete int.

angelayi replied:
Wait, this path is also run when tracing, not just eager. If users constrain a value that is traced to be a static integer this path will also be taken.

XuanQi replied:
I feel like naming is basically the hardest part of programming in general :p. Do we want to make a decision between `constrain_on_int ` and `check_range_and_create_symbol `? cc @angelayi , @avikchaudhuri , @tugsbayasgalan, thoughts?

I kind of leaning towards `constrain_on_int` and will add comments in the function to describe more.

avikchaudhuri replied:
Fine by me!

tugsbayasgalan commented on test/dynamo/test_export.py:
```diff
@@ -2411,6 +2411,29 @@ def f(x, y):
+            def f(x):
+                a = x.item()
+                constrain_as_size(a, 4, 7)
+                return torch.empty((a, 4))
```
Could you add another test where we actually use `torch.empty((a, 4))` somewhere non-trivially? Just wanted to make sure the symbols are propagating properly.

tugsbayasgalan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -177,6 +177,10 @@
 
+- func: sym_constrain_range(Scalar size, int? min=None, int? max=None) -> ()
+  dispatch:
+    CPU: sym_constrain_range_cpu
```
Can you also add dummy CUDA implementation for it?

XuanQi replied:
Will do

eellison commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -439,6 +439,14 @@ def resize_as_(fake_mode, func, *args, **kwargs):
 
+# Don't default to default device handling since range check is only applied to
+# scalar and doing the check as side effect.
+@register_op_impl(aten.sym_constrain_range.default)
```
I think this is unnecessary. What happens if you remove this ?

XuanQi replied:
If I removed it, it will crash at https://github.com/pytorch/pytorch/blob/e82616d9001c6cc7a533b92eaab0b41eba5533e6/torch/_subclasses/fake_tensor.py#L1412 and complains "Could not find common device for aten.sym_constrain_range.default".

My gut feeling is that for our use case this function shouldn't run at all because `sym_constrain_range` returns `None` so it doesn't make sense to infer device for it?

eellison replied:
Im surprised you get this error because there is no Tensor output, right, so we shouldn't be calling the wrap function.

XuanQi replied:
@eellison: So this change looks OK then?

XuanQi replied:
@eellison: Awesome! It works now. Thanks for the help!

angelayi commented on test/dynamo/test_export.py:
```diff
@@ -2411,6 +2411,29 @@ def f(x, y):
+                a = x.item()
+                constrain_as_size(a, 4, 7)
+                return torch.empty((a, 4))
+
```
Could we also add a check using FileCheck (ex. https://fburl.com/code/4exbqxbl) to make sure the op is in the graph?

XuanQi committed:
Update on "[RFC]: Create aten native op for constrain_range"


At high current implementation of constrains functions (constrain_as_**) will raise exception for the following code snippets:
```
def f(x):
    a = x.item()
    constrain_as_size(a, 4, 7)
    return torch.empty((a, 4))

inp = torch.tensor([5])
ep = torch._export.export(f, (inp,))
```

The reason is because current constrain logic is:
1) Purely python so it won't survive AOT export (the full node is gone after AOT export since AOT export only maintains aten level op).
2) Utilize side effect to add range constraints for traced symbol's shape env ([code](https://github.com/pytorch/pytorch/blob/9591e52880fb09cb6753d97606e6efb956075f5b/torch/fx/experimental/symbolic_shapes.py#L370-L372)).
3) If runtime assertion is turned on (by default). [`_AddRuntimeAssertionsForConstraintsPass`](https://github.com/pytorch/pytorch/blob/9591e52880fb09cb6753d97606e6efb956075f5b/torch/_export/passes/add_runtime_assertions_for_constraints_pass.py#L98-L100) will try to append assertion node based on range constrains extracted from shape env of symbol during another interpretation round.
4). However, since 1), in the round of AOT export, range constraints logic won't run for symbols generated during this round. And later there is no range constrains information available for assertion round and caused issue.
5) As a result of above, it will failure at `torch.empty((a, 4))` (there is no constrains for `a` that it must be positive).

The fix here is just to implement range constrain logic as a native aten op (CPU implementation as no-op) to make it be able to survive AOT export.

**NOTE:**
[Logic](https://github.com/pytorch/pytorch/blob/2d745b95d723641e575027bd4e2fff612f61cc8f/torch/fx/experimental/symbolic_shapes.py#L350-L365C15) within [`constrain_range`](https://github.com/pytorch/pytorch/blob/2d745b95d723641e575027bd4e2fff612f61cc8f/torch/fx/experimental/symbolic_shapes.py#LL313C74-L313C74) is split out as `constrain_range_eager` to capture case when non `SymInt` is passed in and reused in the new `_constrain_range`. The reason is when non `SymInt` is provided:
* If it directly calls `sym_constrain_range`, the C++ version will be called which will be no-op.
* So in this case it calls `constrain_range_eager` instead to be able to capture issue like user provides a input whose tensor's shape could be out of range during exporting, like the following for above code example:
```
...
inp = torch.tensor([10])
ep = torch._export.export(f, (inp,)) # immediately raise error
```

Differential Revision: [D46734204](https://our.internmc.facebook.com/intern/diff/D46734204)

[ghstack-poisoned]

XuanQi commented:
@xuanqi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D46734204).

XuanQi committed:
Update on "[RFC]: Create aten native op for constrain_range"


At high current implementation of constrains functions (constrain_as_**) will raise exception for the following code snippets:
```
def f(x):
    a = x.item()
    constrain_as_size(a, 4, 7)
    return torch.empty((a, 4))

inp = torch.tensor([5])
ep = torch._export.export(f, (inp,))
```

The reason is because current constrain logic is:
1) Purely python so it won't survive AOT export (the full node is gone after AOT export since AOT export only maintains aten level op).
2) Utilize side effect to add range constraints for traced symbol's shape env ([code](https://github.com/pytorch/pytorch/blob/9591e52880fb09cb6753d97606e6efb956075f5b/torch/fx/experimental/symbolic_shapes.py#L370-L372)).
3) If runtime assertion is turned on (by default). [`_AddRuntimeAssertionsForConstraintsPass`](https://github.com/pytorch/pytorch/blob/9591e52880fb09cb6753d97606e6efb956075f5b/torch/_export/passes/add_runtime_assertions_for_constraints_pass.py#L98-L100) will try to append assertion node based on range constrains extracted from shape env of symbol during another interpretation round.
4). However, since 1), in the round of AOT export, range constraints logic won't run for symbols generated during this round. And later there is no range constrains information available for assertion round and caused issue.
5) As a result of above, it will failure at `torch.empty((a, 4))` (there is no constrains for `a` that it must be positive).

The fix here is just to implement range constrain logic as a native aten op (CPU implementation as no-op) to make it be able to survive AOT export.

**NOTE:**
[Logic](https://github.com/pytorch/pytorch/blob/2d745b95d723641e575027bd4e2fff612f61cc8f/torch/fx/experimental/symbolic_shapes.py#L350-L365C15) within [`constrain_range`](https://github.com/pytorch/pytorch/blob/2d745b95d723641e575027bd4e2fff612f61cc8f/torch/fx/experimental/symbolic_shapes.py#LL313C74-L313C74) is split out as `constrain_range_int` to capture case when non `SymInt` is passed in and reused in the new `_constrain_range`. The reason is when non `SymInt` is provided:
* If it directly calls `sym_constrain_range`, the C++ version will be called which will be no-op.
* So in this case it calls `constrain_range_int` instead to be able to capture issue like user provides a input whose tensor's shape could be out of range during exporting, like the following for above code example:
```
...
inp = torch.tensor([10])
ep = torch._export.export(f, (inp,)) # immediately raise error
```

Differential Revision: [D46734204](https://our.internmc.facebook.com/intern/diff/D46734204)

[ghstack-poisoned]

XuanQi commented:
@xuanqi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D46734204).

XuanQi committed:
Update on "[RFC]: Create aten native op for constrain_range"


At high current implementation of constrains functions (constrain_as_**) will raise exception for the following code snippets:
```
def f(x):
    a = x.item()
    constrain_as_size(a, 4, 7)
    return torch.empty((a, 4))

inp = torch.tensor([5])
ep = torch._export.export(f, (inp,))
```

The reason is because current constrain logic is:
1) Purely python so it won't survive AOT export (the full node is gone after AOT export since AOT export only maintains aten level op).
2) Utilize side effect to add range constraints for traced symbol's shape env ([code](https://github.com/pytorch/pytorch/blob/9591e52880fb09cb6753d97606e6efb956075f5b/torch/fx/experimental/symbolic_shapes.py#L370-L372)).
3) If runtime assertion is turned on (by default). [`_AddRuntimeAssertionsForConstraintsPass`](https://github.com/pytorch/pytorch/blob/9591e52880fb09cb6753d97606e6efb956075f5b/torch/_export/passes/add_runtime_assertions_for_constraints_pass.py#L98-L100) will try to append assertion node based on range constrains extracted from shape env of symbol during another interpretation round.
4). However, since 1), in the round of AOT export, range constraints logic won't run for symbols generated during this round. And later there is no range constrains information available for assertion round and caused issue.
5) As a result of above, it will failure at `torch.empty((a, 4))` (there is no constrains for `a` that it must be positive).

The fix here is just to implement range constrain logic as a native aten op (CPU implementation as no-op) to make it be able to survive AOT export.

**NOTE:**
[Logic](https://github.com/pytorch/pytorch/blob/2d745b95d723641e575027bd4e2fff612f61cc8f/torch/fx/experimental/symbolic_shapes.py#L350-L365C15) within [`constrain_range`](https://github.com/pytorch/pytorch/blob/2d745b95d723641e575027bd4e2fff612f61cc8f/torch/fx/experimental/symbolic_shapes.py#LL313C74-L313C74) is split out as `constrain_range_int` to capture case when non `SymInt` is passed in and reused in the new `_constrain_range`. The reason is when non `SymInt` is provided:
* If it directly calls `sym_constrain_range`, the C++ version will be called which will be no-op.
* So in this case it calls `constrain_range_int` instead to be able to capture issue like user provides a input whose tensor's shape could be out of range during exporting, like the following for above code example:
```
...
inp = torch.tensor([10])
ep = torch._export.export(f, (inp,)) # immediately raise error
```

Differential Revision: [D46734204](https://our.internmc.facebook.com/intern/diff/D46734204)

[ghstack-poisoned]

XuanQi commented:
@xuanqi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D46734204).

XuanQi committed:
Update on "[RFC]: Create aten native op for constrain_range"


At high current implementation of constrains functions (constrain_as_**) will raise exception for the following code snippets:
```
def f(x):
    a = x.item()
    constrain_as_size(a, 4, 7)
    return torch.empty((a, 4))

inp = torch.tensor([5])
ep = torch._export.export(f, (inp,))
```

The reason is because current constrain logic is:
1) Purely python so it won't survive AOT export (the full node is gone after AOT export since AOT export only maintains aten level op).
2) Utilize side effect to add range constraints for traced symbol's shape env ([code](https://github.com/pytorch/pytorch/blob/9591e52880fb09cb6753d97606e6efb956075f5b/torch/fx/experimental/symbolic_shapes.py#L370-L372)).
3) If runtime assertion is turned on (by default). [`_AddRuntimeAssertionsForConstraintsPass`](https://github.com/pytorch/pytorch/blob/9591e52880fb09cb6753d97606e6efb956075f5b/torch/_export/passes/add_runtime_assertions_for_constraints_pass.py#L98-L100) will try to append assertion node based on range constrains extracted from shape env of symbol during another interpretation round.
4). However, since 1), in the round of AOT export, range constraints logic won't run for symbols generated during this round. And later there is no range constrains information available for assertion round and caused issue.
5) As a result of above, it will failure at `torch.empty((a, 4))` (there is no constrains for `a` that it must be positive).

The fix here is just to implement range constrain logic as a native aten op (CPU implementation as no-op) to make it be able to survive AOT export.

**NOTE:**
[Logic](https://github.com/pytorch/pytorch/blob/2d745b95d723641e575027bd4e2fff612f61cc8f/torch/fx/experimental/symbolic_shapes.py#L350-L365C15) within [`constrain_range`](https://github.com/pytorch/pytorch/blob/2d745b95d723641e575027bd4e2fff612f61cc8f/torch/fx/experimental/symbolic_shapes.py#LL313C74-L313C74) is split out as `constrain_range_int` to capture case when non `SymInt` is passed in and reused in the new `_constrain_range`. The reason is when non `SymInt` is provided:
* If it directly calls `sym_constrain_range`, the C++ version will be called which will be no-op.
* So in this case it calls `constrain_range_int` instead to be able to capture issue like user provides a input whose tensor's shape could be out of range during exporting, like the following for above code example:
```
...
inp = torch.tensor([10])
ep = torch._export.export(f, (inp,)) # immediately raise error
```

Differential Revision: [D46734204](https://our.internmc.facebook.com/intern/diff/D46734204)

[ghstack-poisoned]

XuanQi commented:
@xuanqi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D46734204).

avikchaudhuri reviewed (commented):
Thanks @XuanQi for addressing my feedback. I'm going to defer to the other reviewers to approve.

XuanQi committed:
Update on "[RFC]: Create aten native op for constrain_range"


At high current implementation of constrains functions (constrain_as_**) will raise exception for the following code snippets:
```
def f(x):
    a = x.item()
    constrain_as_size(a, 4, 7)
    return torch.empty((a, 4))

inp = torch.tensor([5])
ep = torch._export.export(f, (inp,))
```

The reason is because current constrain logic is:
1) Purely python so it won't survive AOT export (the full node is gone after AOT export since AOT export only maintains aten level op).
2) Utilize side effect to add range constraints for traced symbol's shape env ([code](https://github.com/pytorch/pytorch/blob/9591e52880fb09cb6753d97606e6efb956075f5b/torch/fx/experimental/symbolic_shapes.py#L370-L372)).
3) If runtime assertion is turned on (by default). [`_AddRuntimeAssertionsForConstraintsPass`](https://github.com/pytorch/pytorch/blob/9591e52880fb09cb6753d97606e6efb956075f5b/torch/_export/passes/add_runtime_assertions_for_constraints_pass.py#L98-L100) will try to append assertion node based on range constrains extracted from shape env of symbol during another interpretation round.
4). However, since 1), in the round of AOT export, range constraints logic won't run for symbols generated during this round. And later there is no range constrains information available for assertion round and caused issue.
5) As a result of above, it will failure at `torch.empty((a, 4))` (there is no constrains for `a` that it must be positive).

The fix here is just to implement range constrain logic as a native aten op (CPU implementation as no-op) to make it be able to survive AOT export.

**NOTE:**
[Logic](https://github.com/pytorch/pytorch/blob/2d745b95d723641e575027bd4e2fff612f61cc8f/torch/fx/experimental/symbolic_shapes.py#L350-L365C15) within [`constrain_range`](https://github.com/pytorch/pytorch/blob/2d745b95d723641e575027bd4e2fff612f61cc8f/torch/fx/experimental/symbolic_shapes.py#LL313C74-L313C74) is split out as `constrain_range_int` to capture case when non `SymInt` is passed in and reused in the new `_constrain_range`. The reason is when non `SymInt` is provided:
* If it directly calls `sym_constrain_range`, the C++ version will be called which will be no-op.
* So in this case it calls `constrain_range_int` instead to be able to capture issue like user provides a input whose tensor's shape could be out of range during exporting, like the following for above code example:
```
...
inp = torch.tensor([10])
ep = torch._export.export(f, (inp,)) # immediately raise error
```

Differential Revision: [D46734204](https://our.internmc.facebook.com/intern/diff/D46734204)

[ghstack-poisoned]

XuanQi commented:
@xuanqi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D46734204).

XuanQi commented:
@xuanqi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D46734204).

XuanQi committed:
Update on "[RFC]: Create aten native op for constrain_range"


At high current implementation of constrains functions (constrain_as_**) will raise exception for the following code snippets:
```
def f(x):
    a = x.item()
    constrain_as_size(a, 4, 7)
    return torch.empty((a, 4))

inp = torch.tensor([5])
ep = torch._export.export(f, (inp,))
```

The reason is because current constrain logic is:
1) Purely python so it won't survive AOT export (the full node is gone after AOT export since AOT export only maintains aten level op).
2) Utilize side effect to add range constraints for traced symbol's shape env ([code](https://github.com/pytorch/pytorch/blob/9591e52880fb09cb6753d97606e6efb956075f5b/torch/fx/experimental/symbolic_shapes.py#L370-L372)).
3) If runtime assertion is turned on (by default). [`_AddRuntimeAssertionsForConstraintsPass`](https://github.com/pytorch/pytorch/blob/9591e52880fb09cb6753d97606e6efb956075f5b/torch/_export/passes/add_runtime_assertions_for_constraints_pass.py#L98-L100) will try to append assertion node based on range constrains extracted from shape env of symbol during another interpretation round.
4). However, since 1), in the round of AOT export, range constraints logic won't run for symbols generated during this round. And later there is no range constrains information available for assertion round and caused issue.
5) As a result of above, it will failure at `torch.empty((a, 4))` (there is no constrains for `a` that it must be positive).

The fix here is just to implement range constrain logic as a native aten op (CPU implementation as no-op) to make it be able to survive AOT export.

**NOTE:**
[Logic](https://github.com/pytorch/pytorch/blob/2d745b95d723641e575027bd4e2fff612f61cc8f/torch/fx/experimental/symbolic_shapes.py#L350-L365C15) within [`constrain_range`](https://github.com/pytorch/pytorch/blob/2d745b95d723641e575027bd4e2fff612f61cc8f/torch/fx/experimental/symbolic_shapes.py#LL313C74-L313C74) is split out as `constrain_range_int` to capture case when non `SymInt` is passed in and reused in the new `_constrain_range`. The reason is when non `SymInt` is provided:
* If it directly calls `sym_constrain_range`, the C++ version will be called which will be no-op.
* So in this case it calls `constrain_range_int` instead to be able to capture issue like user provides a input whose tensor's shape could be out of range during exporting, like the following for above code example:
```
...
inp = torch.tensor([10])
ep = torch._export.export(f, (inp,)) # immediately raise error
```

Differential Revision: [D46734204](https://our.internmc.facebook.com/intern/diff/D46734204)

[ghstack-poisoned]

XuanQi commented:
@xuanqi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D46734204).

XuanQi commented:
@pytorchbot rebase

XuanQi committed:
Update on "[RFC]: Create aten native op for constrain_range"


At high current implementation of constrains functions (constrain_as_**) will raise exception for the following code snippets:
```
def f(x):
    a = x.item()
    constrain_as_size(a, 4, 7)
    return torch.empty((a, 4))

inp = torch.tensor([5])
ep = torch._export.export(f, (inp,))
```

The reason is because current constrain logic is:
1) Purely python so it won't survive AOT export (the full node is gone after AOT export since AOT export only maintains aten level op).
2) Utilize side effect to add range constraints for traced symbol's shape env ([code](https://github.com/pytorch/pytorch/blob/9591e52880fb09cb6753d97606e6efb956075f5b/torch/fx/experimental/symbolic_shapes.py#L370-L372)).
3) If runtime assertion is turned on (by default). [`_AddRuntimeAssertionsForConstraintsPass`](https://github.com/pytorch/pytorch/blob/9591e52880fb09cb6753d97606e6efb956075f5b/torch/_export/passes/add_runtime_assertions_for_constraints_pass.py#L98-L100) will try to append assertion node based on range constrains extracted from shape env of symbol during another interpretation round.
4). However, since 1), in the round of AOT export, range constraints logic won't run for symbols generated during this round. And later there is no range constrains information available for assertion round and caused issue.
5) As a result of above, it will failure at `torch.empty((a, 4))` (there is no constrains for `a` that it must be positive).

The fix here is just to implement range constrain logic as a native aten op (CPU implementation as no-op) to make it be able to survive AOT export.

**NOTE:**
[Logic](https://github.com/pytorch/pytorch/blob/2d745b95d723641e575027bd4e2fff612f61cc8f/torch/fx/experimental/symbolic_shapes.py#L350-L365C15) within [`constrain_range`](https://github.com/pytorch/pytorch/blob/2d745b95d723641e575027bd4e2fff612f61cc8f/torch/fx/experimental/symbolic_shapes.py#LL313C74-L313C74) is split out as `constrain_range_int` to capture case when non `SymInt` is passed in and reused in the new `_constrain_range`. The reason is when non `SymInt` is provided:
* If it directly calls `sym_constrain_range`, the C++ version will be called which will be no-op.
* So in this case it calls `constrain_range_int` instead to be able to capture issue like user provides a input whose tensor's shape could be out of range during exporting, like the following for above code example:
```
...
inp = torch.tensor([10])
ep = torch._export.export(f, (inp,)) # immediately raise error
```

Differential Revision: [D46734204](https://our.internmc.facebook.com/intern/diff/D46734204)

[ghstack-poisoned]

XuanQi commented:
@xuanqi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D46734204).
